File: 1101-1200/1120-Maximum Average Subtree/1120-Maximum Average Subtree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def maximumAverageSubtree(self, root: TreeNode) -> float:
        maxAvg = 0
        def dfs(root):
            if root is None:
                return 0, 0
            nonlocal maxAvg
            leftSum, leftCount = dfs(root.left)
            rightSum, rightCount = dfs(root.right)
            s = leftSum + root.val + rightSum
            count = leftCount + 1 + rightCount
            maxAvg = max(maxAvg, s / count)
            return s, count
        dfs(root)
        return maxAvg

File: 1101-1200/1120-Maximum Average Subtree/1120-Maximum Average Subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    double maximumAverageSubtree(TreeNode* root) {
        double maxAvg = 0;
        dfs(root, maxAvg);
        return maxAvg;
    }
    
private:
    pair<int, int> dfs(TreeNode* root, double& maxAvg) {
        if (root == nullptr) {
            return {0, 0};
        }
        auto [leftSum, leftCount] = dfs(root->left, maxAvg);
        auto [rightSum, rightCount] = dfs(root->right, maxAvg);
        int sum = leftSum + root->val + rightSum;
        int count = leftCount + 1 + rightCount;
        maxAvg = max(maxAvg, (double)sum / count);
        return {sum, count};
    }
};
